edb/schema/links.py

```diff
@@ -362,12 +362,38 @@ def _apply_field_ast(
                         assert isinstance(t, (so.Object, so.ObjectShell))
                         node.target = utils.typeref_to_ast(schema, t)
             else:
+                old_type = pointers.merge_target(
+                    self.scls,
+                    list(self.scls.get_bases(schema).objects(schema)),
+                    'target',
+                    ignore_local=True,
+                    schema=schema,
+                )
                 assert isinstance(op.new_value, (so.Object, so.ObjectShell))
+                new_type = (
+                    op.new_value.resolve(schema)
+                    if isinstance(op.new_value, so.ObjectShell)
+                    else op.new_value)
+
+                new_type_ast = utils.typeref_to_ast(schema, op.new_value)
+                cast_expr = None
+                # If the type isn't assignment castable, generate a
+                # USING with a nonsense cast. It shouldn't matter,
+                # since there should be no data to cast, but the DDL side
+                # of things doesn't know that since the command is split up.
+                if old_type and not old_type.assignment_castable_to(
+                        new_type, schema):
+                    cast_expr = qlast.TypeCast(
```

Comment:
Hmm. I'm slightly surprised this works, since type casts from a known object type to another object type are ought to be illegal:

```
edgedb> CREATE TYPE Foo;
OK: CREATE TYPE
edgedb> CREATE TYPE Bar;
OK: CREATE TYPE
edgedb> SELECT <Bar>Foo;
error: QueryError: cannot cast object type 'default::Foo' to 'default::Bar', use `...[IS default::Bar]` instead
  ┌─ query:1:13
  │
1 │ SELECT <Bar>Foo;
  │             ^^^ error
```

Reply:
Empty set. OK